### DISCUSSION TITLE
Is any way to import openai?

### DISCUSSION THREAD

**Discussion by wpsoul** [pyscript/pyscript#1415] [Q&A]
I am trying to import openai with

import openai

but ModuleNotFoundError: No module named 'openai'

Tried on local machine and on pyscript.com

**Comment by FabioRosado** [comment]
Hello, how are you installing openai?

Pyscript doesn't look at your local files/dependencies since it's running in the browser so the environment is different. I have not tried to install openai with PyScript but you will need to add the following configuration to your index HTML file:

```html
<py-config>
  packages = ["openai"]
</py-config>
```

Now you may see the following error in the page: `Couldn't find a pure Python 3 wheel for <dependency>` which means that openai or any dependency used with it needs to be ported to work in WASM and on PyScript.

Hope this helps

**Comment by wpsoul** [comment]
Yes, this is what I have done.

`<py-config>
  packages = ["openai"]
</py-config>`

And I get next error

(PY1001): Unable to install package(s) 'openai'. Reason: Can't find a pure Python 3 Wheel for package(s) 'openai'. See: https://pyodide.org/en/stable/usage/faq.html#micropip-can-t-find-a-pure-python-wheel for more information.

But I can't see the way to port it or documentation about it

**Reply by hoodmane** [reply]
Porting OpenAI is most likely a very heavy lift. If you are not very experienced with build systems and unusual compilers and linkers, it is not likely that you will be able to make progress.

**Reply by wpsoul** [reply]
But why I can import other libraries via py-config? I thought it should be easy to do the same with openai. I did it in Streamlit in few seconds. But pyscript gives me much more freedom in UI and interactions between javascript, HTML and python. It should have openai, langchain, because this will be most popular library for Pyscript

**Reply by JeffersGlass** [reply]
As far as I know, Streamlit is running all your Python server side, which is where Python packages 'expect' to be run. By contrast, PyScript/Pyodide run in the browser, which means (a) any non-Python extensions need to be pre-compiled to Web Assembly and (b) some resources, like low-level access to networking functions, threading, and the GPU aren't available because of limitations running inside the browser window.

**Reply by hoodmane** [reply]
Some people are looking into these libraries but unfortunately many of them have very complicated build systems and require a very large amount of effort. Unfortunately there is currently nobody in the world who both knows how to port these libraries and is willing and able to devote the amount of time it requires. It is difficult to know how many different issues will arise when you start. If you spend a week fixing the thing that currently breaks the build, the next place the build breaks may be equally complicated. So it is very hard to gauge progress on these tasks.

For reference, here is my blog post about what it took to port the cryptography package to Pyodide:
https://blog.pyodide.org/posts/rust-pyo3-support-in-pyodide/
This required months of work, learning various arcane topics about linkers, patching upstream typos, pull requests to patch the rust compiler, fixes to pyo3 and maturin, etc etc. I did this because (1) it was a very frequently requested feature and (2) I had personal interest in Rust packages.

There is reason to think OpenAI might be a bit easier, but my guess is it would take a very experienced dev at least a month of focused work.

**Reply by wpsoul** [reply]
Clear, then, I will stick with JS API from OpenAi
